### DISCUSSION TITLE
Installing Gleam, Erlang, and Elixir

### DISCUSSION THREAD

**Discussion by lpil** [gleam-lang/gleam#3892] [Ideas & suggestions]
It would be useful to have a tool to install these dependencies. The main challenge is where do we get the builds from?

- Gleam can be taken from gleam-lang/gleam
- Elixir can be taken from elixir-lang/elixir
- Erlang
  - Windows can be taken from erlang/otp
  - macOS can be taken from erlef/otp_builds
  - Linux has hexpm/bob, but it's **Ubuntu only**.

We need something for Linux.

Related: https://github.com/erlef/otp_builds/issues/39

**Comment by eshikafe** [comment]
What about using [kerl](https://github.com/kerl/kerl) to compile Erlang on any Linux distribution (on the host machine)?

**Reply by lpil** [reply]
We would not be able to manage the complexity of Kerl sufficiently to make it part of the Gleam toolchain officially.

**Reply by eshikafe** [reply]
What about using `asdf` so you don't have to worry about `kerl`?

**Reply by lpil** [reply]
asdf wraps kerl and does not remove its challenges. Making two tools part of the official toolchain is more complex, not less complex.

**Reply by eshikafe** [reply]
asdf has pre-built binaries similar to Gleam. Gleam only needs to interface with asdf to install Erlang. Anyway, I understand your point about having two different tools involved. This can create problems in the long run.

**Reply by lpil** [reply]
Please see the original post about the available builds.
asdf does not solve the problem here ad they do not have the missing build.

Implementing a tool like asdf is trivial and exactly what tool we use is not something we are concerned with today.

**Reply by tamadamas** [reply]
> asdf has pre-built binaries similar to Gleam. Gleam only needs to interface with asdf to install Erlang. Anyway, I understand your point about having two different tools involved. This can create problems in the long run.

Asdf? Please, whatever but not asdf, just check these security guides https://mise.jdx.dev/dev-tools/comparison-to-asdf.html#supply-chain-security

Asdf uses bash for plugins even if core is rewritten to `Go`

**Reply by lpil** [reply]
Again, we are not concerned with selecting a front-end tool such as asdf at this point. That aspect is trivial, the missing Linux build is the concern today.

**Reply by tamadamas** [reply]
> Again, we are not concerned with selecting a front-end tool such as asdf at this point. That aspect is trivial, the missing Linux build is the concern today.

@lpil Have you read that my comment? 
https://github.com/gleam-lang/gleam/discussions/3892#discussioncomment-12512569

I'm exactly about Linux build. That had worried me within Elixir dev days also

**Comment by tamadamas** [comment]
We already have enough tools.
My choice is `mise`. It wraps aqua, vfox, ubi, asdf.

Also you can check any of those above directly if they match your needs

**Reply by GearsDatapacks** [reply]
The idea is to make the installation process of Gleam easier and simpler, without having to install external tools

**Reply by lpil** [reply]
And none of those tools are good enough. This discussion is to address the problems we have because of the shortcomings of them and OS package managers.

**Reply by eshikafe** [reply]
@lpil what's the ideal solution you are looking for? Let's see how we can help achieve it.

**Reply by lpil** [reply]
The first step is to get a generic Linux build of Erlang/OTP. Once we have that the rest becomes straightforward.

**Reply by tamadamas** [reply]
Could we use Zig for building Erlang?
I pretty much loving how they work on building C-related code

**Reply by lpil** [reply]
I don't have any particular opinions on what C compiler would work best. It'd be up to whoever maintains the build as to what approach they wish to use.

**Reply by tamadamas** [reply]
> I don't have any particular opinions on what C compiler would work best. It'd be up to whoever maintains the build as to what approach they wish to use.

Zig is not just a "c compiler", it's a new language(like rust) and build toolchain (build.zig) with that language that can compile C/c++/zig with different arch/os without any makefiles.

Check this for ffmpeg for example 
https://github.com/allyourcodebase/ffmpeg/blob/main/build.zig

https://github.com/allyourcodebase

I feel there is a solution.
I would be one of contributors because anyway I'm learning Zig and Gleam right now(Erlang and Elixir => all in my heart already)

**Reply by lpil** [reply]
I am familiar with the Zig C compiler, yes.

It is up to the maintainers of the build to decide what is the best approach for them to use.

**Reply by tamadamas** [reply]
> I am familiar with the Zig C compiler, yes.
> 
> It is up to the maintainers of the build to decide what is the best approach for them to use.

Sure, making amazing easy Erlang building system is a great team work for all BEAM community if we want to. And it's great we're discussing about it.

**Reply by mine-tech-oficial** [reply]
On the topic, maybe the dev behind `mise` might want to help, as Erlang is one of mise's core tools

**Reply by lpil** [reply]
asdf and mise don't provide any builds at all, so it's unlikely they'd be positioned to help with this C build process development.

**Reply by mine-tech-oficial** [reply]
Good point

**Reply by tamadamas** [reply]
> On the topic, maybe the dev behind `mise` might want to help, as Erlang is one of mise's core tools

Are you kidding, right? Everything they work for now as I can see, it's adding proxy to aqua/ubi registries. It's good but it's carrying water from left to the right.
The zig community is the real thing here.
Even `rust` uses makefiles https://github.com/rust-lang/rust/blob/master/INSTALL.md#configure-and-make
And... python...WAT? ^_0

**Reply by eshikafe** [reply]
> Could we use Zig for building Erlang? I pretty much loving how they work on building C-related code

This will help simplify creating a generic build for Linux. However, a new repo would have to be created and maintained specifically for this. We can collaborate on this @tamadamas if it's something you'd like to work on. This will definitely solve the problem of this discussion.

**Reply by tamadamas** [reply]
> > Could we use Zig for building Erlang? I pretty much loving how they work on building C-related code
> 
> This will help simplify creating a generic build for Linux. However, a new repo would have to be created and maintained specifically for this. We can collaborate on this @tamadamas if it's something you'd like to work on. This will definitely solve the problem of this discussion.

Sure 😊

**Reply by mine-tech-oficial** [reply]
> > On the topic, maybe the dev behind `mise` might want to help, as Erlang is one of mise's core tools
> 
> Are you kidding, right? Everything they work for now as I can see, it's adding proxy to aqua/ubi registries. It's good but it's carrying water from left to the right. The zig community is the real thing here. Even `rust` uses makefiles https://github.com/rust-lang/rust/blob/master/INSTALL.md#configure-and-make And... python...WAT? ^_0

I thought that, because Erlang was one of their core tools, they'd like if there were more prebuilt binaries, that's why I commented.

**Comment by tamadamas** [comment]
So I've started with https://github.com/exfu/erlang.zig

> Could we just do `zig build` to make Erlang
>
> Could we just do `zig build --with elixir --with gleam`
>
> On Linux, MacOS, Windows, WhateverOS ^_^
>
© Don't force me to read any Manual, please

**Reply by eshikafe** [reply]
So, what would be an appropriate name for the generic Linux build?

To be consistent with the Erlang/OTP releases for Windows, I think the following should be fine: otp_linux-{arch}_{OTP-Release}.tar.gz e.g `otp_linux-x86_64_27.3.tar.gz`, `otp_linux-x86-64_27.3.tar.gz` or `otp_linux-amd64_27.3.tar.gz` for 64-bit, and `otp_linux-x86_27.3.tar.gz` for 32-bit.


![image](https://github.com/user-attachments/assets/cf050d6c-6353-4018-ac4b-1e7586903e2a)

@tamadamas: Please, add a `Discussion` section to https://github.com/exfu/erlang.zig.
